### PR TITLE
KAFKA-7131 Update release script to generate announcement email text

### PR DESCRIPTION
**Pull request by bibinss**

### CODE REVIEW

**Comment by bibinss** [comment]
@mjsax can you please take a look?

File: release.py
Comment (mjsax) [inline release.py]: If we run the script to do the actual release, we have this information already. It would be good to reuse this. Ie, we can keep this as-is, however add a second method that takes this information as parameters. This allow us to call the new method from here, after we collected the information, but also at the end of the regular execution of the script and pass in the information directly. Thus, if a committer does a release, it's not required to call the script again but the email template will be generated directly.

**Review by mjsax — COMMENTED** [review]
Sorry for late review and thanks for the PR. One initial comment. Did you test the script?

\cc @ewencp for second review

File: release.py
Comment (bibinss) [reply release.py]: @mjsax couple of questions
1 - "If we run the script to do the actual release, we have this information already" - 
Do you mean to say that we can get the current and previos release versions from git tags? like first fetch all the tags, sort in descending order and pick first and second indexed values are current and previos versions?

2 - "This allow us to call the new method from here, after we collected the information, but also at the end of the regular execution of the script and pass in the information directly. Thus, if a committer does a release, it's not required to call the script again but the email template will be generated directly" - 
The script 'release.py' if run, by default generates the email content for RC release for voting. At this point we are not ready for sending release completion announcement. So, what is point of generating the announcement email content at the end of the regular execution of the script 'release.py'?
I mean, we need to send release announcement email only once voting is approved and we manually tag the release branch last commit with release version.

Comment (ewencp) [reply release.py]: You definitely can determine this automatically from the existing tags. For anything with patch version > 0, it's trivial since you want the reference for previous version to be `patch_version - 1`. For the first release in a major.minor release line, you would need to figure out the correct previous major.minor.patch release and use that.

Normally I would say this is pretty easy, just list tags, find ones that match the right pattern, split segments on `.` characters, convert each to integers, and sort. However, this does get a bit messier with Kafka because of the switch in release numbering (from 4 digits in pre-1.0 to 3 digits in post-1.0), so you'd have to normalize to 4 digits, sort, then make sure you drop any extra digits from post-1.0 versions.

It'd be nice to get this all automated and the ergonomics of the script are nicer if they it is, but I wouldn't block merging this on that. This is still better than what committers do today, which is to just construct this all manually.

Comment (ewencp) [inline release.py]: nit: `previos` -> `previous`

File: release.py
Comment (ewencp) [inline release.py]: Should we do validation here like we do in the default staging command, i.e. like this https://github.com/apache/kafka/blob/trunk/release.py#L330-L338? Maybe we can refactor the validation/setting of global variable `release_version_parts` there so we can just reuse that bit of code?

File: release.py
Comment (ewencp) [inline release.py]: This should go to `announce@apache.org` as well, that's actually the most critical one as that makes it "official".

File: release.py
Comment (ewencp) [inline release.py:380]: might want to also remind them that the email *must* come from their apache email address so it will be correctly delivered to `announce@apache.org` (see earlier comment). i.e. addressing this part of the wiki text that you are automating:

> You will need to use your apache email address to send out the email (otherwise, it won't be delivered to announce@apache.org)

File: release.py
Comment (ewencp) [inline release.py]: Not sure how critical it is that we automate this, but these are configurable in the build now (see `availableScalaVersions`/`defaultScalaVersions` in `dependencies.gradle`). I'm a bit wary of trying to automate just because it requires parsing gradle code to do robustly, but at the same time, this seems like it'd be easy to have fall out of date.

I guess we could leave as is and hopefully people will be comprehensively `git grep`ing for places they need to change when adding/removing scala versions?

File: release.py
Comment (ewencp) [inline release.py]: This command has always left a trailing `,`. You could potentially omit the commands after the `cut` and just do a split/join in python that will give exactly what we want.

Also, not sure if it was intentional or not, but this command seems to elide the alphabetical sorting that's in the command on the wiki.

**Review by ewencp — COMMENTED** [review]
@bibinss Some details, but this is mostly looking good to me. Thanks for taking this on! I'm a huge fan of anything we can do to further automate the (otherwise somewhat tedious) release process.

File: release.py
Comment (bibinss) [reply release.py:380]: this is done

File: release.py
Comment (bibinss) [reply release.py]: this is done

File: release.py
Comment (bibinss) [reply release.py]: this is done

File: release.py
Comment (bibinss) [reply release.py]: yes, refactored the code

File: release.py
Comment (bibinss) [reply release.py]: yes, done

File: release.py
Comment (bibinss) [reply release.py:380]: yes, done

**Comment by bibinss** [comment]
@ewencp @mjsax I have added the logic to find the current and previous release from git tags. 
Also did work on most of your review comments. can you please take a look at the updated PR?

**Comment by bibinss** [comment]
@ewencp @mjsax ^^

**Comment by bibinss** [comment]
@ewencp  @mjsax can you please take a look at the updated PR?

File: release.py
Comment (mjsax) [reply release.py]: Should be ok to do either 3-digit or 4-digit code (for corresponding branches) ? No need to support both in one branch IMHO

File: release.py
Comment (mjsax) [inline release.py:267]: I like this.

We will be able to pack port this to `2.0`, `1.1`, and `1.0` branch -- not sure if there will ever be a `0.11.0.4` branch -- thus, might not be worth do back port further to support 4-digits.

File: release.py
Comment (mjsax) [reply release.py]: I double people will update this script correctly. We can only hope, that release managers verify this before sending the email...

As an alternative, we can also wildcard this, and let release manger insert those manually. Similar to `<DETAILS OF THE CHANGES>` above.

**Review by mjsax — COMMENTED** [review]
Sorry for the delay. Was quite busy with the feature freeze deadline. Overall LGTM. Some more follow up.

Also: We need some additional cleanup in Line 679ff -- we should also call the new method there automatically.

File: release.py
Comment (ewencp) [reply release.py:267]: we should plan on the possibility of another 0.11 release simply due to possible security fixes.

i agree validating release versions is nice. we can make this handle both old and new version schemes though. pre 1.0, there are 4 version parts but the first version part would always be `0`, so it's pretty easy to handle both cases.

Comment (mjsax) [reply release.py:267]: For an 0.11 release, the committer can still put the email together manually... We need some cutoff point for back porting -- it's just a convenience improvement for doing a release, not a bug fix. Thus, 1.0.0 seems to be a clean cut-off point IMHO.

**Comment by bibinss** [comment]
@mjsax 
- Added a place holder for scala verison in the generated email content and updated PR.
- On this comment " We need some additional cleanup in Line 679ff -- we should also call the new method there automatically." - Are you asking to call new method 'command_release_announcement_email'? why we need to do this at Line 679? at this line we are generating the 'Voting' email for RC right? why do we need to generate release complete email at this point?

**Comment by bibinss** [comment]
@mjsax @ewencp can you please check

**Review by mjsax — APPROVED** [review]
LGTM.

Leave it to @ewencp to have a final look and merge.

File: release.py
Comment (ewencp) [inline release.py:296]: This check isn't quite right since it would just check lexicographically (e.g. `'2.9.0' > '2.11.0'`, but the former is a smaller version than the latter). For now this will probably work fine since we don't hit minor version numbers that large, but eventually this won't work (e.g. major version numbers go > 9). (Similarly, the sort of tags isn't quite right either.)

**Review by ewencp — APPROVED** [review]
LGTM, merging to trunk and back to 1.0

**Comment by mjsax** [comment]
>  and back to 1.0

@ewencp Did you mean 2.0?

**Comment by ewencp** [comment]
Well, I meant 1.0 when I wrote that, but then it didn’t cherrypick cleanly. Left note on the JIRA. In practice, shouldn’t matter since you can use release script from trunk to release older versions.

**Pull request closed without merging**
